### DISCUSSION TITLE
Conditions for Definitions

### DISCUSSION THREAD

**Discussion by cueckoo** [cue-lang/cue#961] [Q&A]
_Originally opened by @embano1 in https://github.com/cuelang/cue/discussions/961_

Hi,

I am not sure if this is possible at all, but here's my use case: given an input `YAML` file I want to validate that if property `type` is set to "A" `aConfig` must be provided (but not `bConfig`) and visa versa. Here's an attempt to show what I'm trying to do but note that the definition does not work:

```
package config

#providerType: "A" | "B"

#provider: {
	type: #providerType
	if #provider.type == "A" {
		aConfig: #aConfig
	}

	if #provider.type == "B" {
		bConfig: #bConfig
	}
}

#aConfig: {
	aAddress: *"https://vc.local/sdk" | string
}

#bConfig: {
	bAddress: *"https://vc.local/sdk" | string
}
```

And given this `yaml` it should fail validation:

```yaml
cfg:
  type: "B"
  aConfig: 
    aAddress: "https://"
```

Is this possible at all?

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/discussions/961#discussioncomment-717780_

You would typically write something like:
```
package config

#providerType: "A" | "B"

#provider: type: #providerType

#provider: {
	type: "A"
	aConfig: #aConfig
} | {
	type: "B"   
	bConfig: #bConfig
}

#aConfig: {
	aAddress: *"https://vc.local/sdk" | string
}

#bConfig: {
	bAddress: *"https://vc.local/sdk" | string
}
```

**Reply by cueckoo** [reply]
_Original reply by @mpvl in https://github.com/cuelang/cue/discussions/961#discussioncomment-717780_

The other schema should work as well though. Which command were you trying to run?

**Reply by cueckoo** [reply]
_Original reply by @mpvl in https://github.com/cuelang/cue/discussions/961#discussioncomment-717780_

Ah, I see. There are two additional issues with the above:

1) The schema file still needs a `cfg: #provider`.

2) `#provider.type == "A"` should be written as `type == "A"`.

`#provider.type` will always resolve to `type` value in the definition. Even in `x & #provider`, the `if` clauses in `#provider` will point to the `type` of `#provider`, not the result of `x & #provider`.

**Reply by cueckoo** [reply]
_Original reply by @embano1 in https://github.com/cuelang/cue/discussions/961#discussioncomment-717780_

Thank you Marcel for your quick response!

In fact I had a `cfg: #provider` and made it work using some hacks as I thought they were :) The issue why this did not work for me in the end was that I could not generate an OpenAPI spec from this, bc it complained about the `cfg` object.

```console
$ openapi: unsupported top-level field "cfg":
    ./config.cue:23:1
```

**Reply by cueckoo** [reply]
_Original reply by @embano1 in https://github.com/cuelang/cue/discussions/961#discussioncomment-717780_

> 2) `#provider.type == "A"` should be written as `type == "A"`.

Ah nice, wasn't sure this would (should) work, bc I thought I have to do full field references inside definitions so they're correctly resolved. This scoping and field name resolution is even better!

**Reply by cueckoo** [reply]
_Original reply by @embano1 in https://github.com/cuelang/cue/discussions/961#discussioncomment-717780_

```
#provider: {
	type: "A"
	aConfig: #aConfig
} | {
	type: "B"   
	bConfig: #bConfig
}
```

If I have multiple type "switches" in this scenario (think K8s `VolumeSource` type as a valid example), this could get messy, right?

**Reply by cueckoo** [reply]
_Original reply by @mpvl in https://github.com/cuelang/cue/discussions/961#discussioncomment-717780_

As long as all types are known and the user is required to specify a type (I believe this is the case for `VolumeSource`, IIRC), this works fine. There are some performance issues with this approach, but they are fixable.

In the case where a constraint is only added for known types, things get a bit tricky. You would then have to resort to the `if` statement as you did. With the query proposal this gets a bit neater again. One could then have a list with all known types and do
```
#provider: {
    T=type: string

    for x in #KnownTypes.[_:{type: T}] { x }  // mix in constraints for known types
}
#KnownTypes: [
    { type: "A", ...},
    { type: "B", ...},
]
```
But this is not yet implemented.

**Reply by cueckoo** [reply]
_Original reply by @mpvl in https://github.com/cuelang/cue/discussions/961#discussioncomment-717780_

Note, btw, that you can also write `#provider: #TypeA | #TypeB | ... | #TypeN` and define the types elsewhere.

**Reply by cueckoo** [reply]
_Original reply by @mpvl in https://github.com/cuelang/cue/discussions/961#discussioncomment-717780_

> $ openapi: unsupported top-level field "cfg":
    ./config.cue:23:1

Interesting point. If you think it will be useful to simply ignore such top-level regular fields, or to allow the `--force` or `-i` flag to override this behavior, feel free to file a feature request.

**Reply by cueckoo** [reply]
_Original reply by @embano1 in https://github.com/cuelang/cue/discussions/961#discussioncomment-717780_

Thank you Marcel, I'll file an issue for this.

Just curious: why does `cue vet` not complain if I omit the `cfg` directive or have it in the schema as `#cfg: ...` definition instead of struct type?

`cue vet` validation with the above yaml example would only fail if I have a `cfg:` struct defined (which then leads to the openapi issue).

I thought having a schema (with definitions only) would suffice when using it in `cue vet` against a yaml file to validate?

May be I'm missing something here.

**Reply by jbcpollak** [reply]
>  If you think it will be useful to simply ignore such top-level regular fields, or to allow the --force or -i flag to override this behavior, feel free to file a feature request.

@mpvl @embano1 - did you ever create a ticket for this and was it ever completed? Could you link it here? I'm doing research on being able to generate OpenAPI schema's from cue files so interested in this topic.

Thanks.

**Comment by AntoineThebaud** [comment]
Hello, I also have a case where I want to declare a definition within a `if` statement and I'm not able to do what I want. Here's my current state:
```
$ cat schema.cue
package pkg

a: string
b: #b
#b: _

$ cat plugin1.cue
package pkg

if a == "a1" {
  #b: {
    c: "c1"
  }
}

$ cat plugin2.cue
package pkg

if a == "a2" {
  #b: {
    c: "c2"
  }
}

$ cat data.cue
package pkg

a: "a1"

$ cue eval
a: "a1"
b: {
    c: "c1"
}
#b: {
    c: "c1"
}
```

So basically here you can imagine that, in my application
- each `if` statement is in a separated file, as a kind of "plugin" that can be added/removed anytime
- the concrete value of `a` is injected dynamically once we receive a new (JSON) payload to be validated.

That's why we use `if` statements here and not a disjunction operator, to allow this dynamicity.

When a new JSON payload is received by the app, we check it with CUE, based on the defs/files available.

The problem with the above solution is that, if ever an unwanted value is set for `a` (let's say `"a3"`), we would like the validation to fail because there's no `#b` def matching. As I said we can not change the `a` declaration like `a: "a1" | "a2"` because of the dynamicity we want to have.

But here in schema.cue we have a kind of "default" constraint `#b: _` that makes any value for `b` to be accepted in this case. We have this because removing it causes this:
```
$ cat schema.cue
package pkg

a: string
b: #b

$ cat plugin1.cue
package pkg

if a == "a1" {
  #b: {
    c: "c1"
  }
}

$ cat plugin2.cue
package pkg

if a == "a2" {
  #b: {
    c: "c2"
  }
}

$ cat data.cue
package pkg

a: "a1"

$ cue eval
b: reference "#b" not found:
    .\file.cue:4:4
```

Even if it is certain here that the first if statement `a == "a1"` will evaluate to true and provide a definition for `#b`, we get this reference not found error. 

So in the end my question is:
- is there a way to make CUE resolve the if statements before trying to resolve the references? If not, is it the kind of feature that could be accepted?
- as an alternative, is there any way to generate a disjunction using a for loop & a map? I don't think that's possible but maybe I missed that..

**Reply by verdverm** [reply]
CUE is intentionally designed to be order independent, so having behavior that depends on order is a bug

You can use the `or([..._])` to create a disjunction from a list

For what you are describing (not sure if it is more complex), I might read the value from the JSON payload into CUE, inspect the value, and then call Unify with the correct definiton, essentially moving the conditionals into the Go. Though I would say disjunctions are probably the more idiomatic approach.

or, put the definitions at the top level and put `b: #b` inside the if? maybe that would cover your use case?

**Reply by myitcv** [reply]
Building on @verdverm's answer, you could do something like this:

```
exec cue eval

-- schema.cue --
package pkg

#as: [string]: _
a: or([for k, _ in #as {k}])
b: #b
#b: _

-- plugin1.cue --
package pkg

#as: a1: true

if a == "a1" {
  #b: {
    c: "c1"
  }
}

-- plugin2.cue --
package pkg

#as: a2: true

if a == "a2" {
  #b: {
    c: "c2"
  }
}

-- data.cue --
package pkg

a: "a3"
```

which (given the `data.cue` file above) gives:

```
> exec cue eval
[stderr]
a: conflicting values "a1" and "a3":
    ./data.cue:3:4
    ./schema.cue:4:8
    ./schema.cue:4:25
a: conflicting values "a2" and "a3":
    ./data.cue:3:4
    ./schema.cue:4:8
    ./schema.cue:4:25
a: 2 errors in empty disjunction::
    ./plugin1.cue:5:4
```

Note with the [query extension](https://github.com/cue-lang/cue/issues/165) the `#as` definition could be transformed into an associative list, which would look and feel a bit better.

**Reply by AntoineThebaud** [reply]
Great, thanks for your quick replies! It seems indeed `or([..._])` was the thing I was looking for.

@verdverm Just to be sure about _CUE is intentionally designed to be order independent, so having behavior that depends on order is a bug_
-> Is there an unwanted behavior in the way the things currently work then? Or is it expected that this kind of schema fails?
```
$ cat file.cue
package pkg

a: "a1"
if a == "a1" {
  #b: "b1"
}
b: #b

$ cue eval
b: reference "#b" not found:
    .\file.cue:7:4
```

**Reply by myitcv** [reply]
@AntoineThebaud no, that's working as intended. Because `#b` is not defined in the outer scope, only the scope define by the `if` comprehension. Hence the reference to `#b` cannot be resolved.
